### DISCUSSION TITLE
changeDynamics fails to set the maxJointVelocity when used in a loop

### DISCUSSION THREAD

**Discussion by gfadini** [bulletphysics/bullet3#3962] [Q&A] [answered]
Hi,
I'm using pybullet to perform some simulations on robotics systems, and I'm pretty impressed by the ease of use and accessibility of the project, so I'd like to thank first of all the developers and the community who made it possible.
Secondly, I'm writing here because I'm facing an issue that maybe can be easily solved because I'm getting the API wrong.
What I'd like to do is to set some maximum joint limits for each joint independently in the loop, from my understanding it should be possible with the function ```changeDynamics``` that I'm using in this way:
`
        max_vel = [0.11586667473156319, 10.065067067830148, 19.193867310528766, 8.134619738240676]

        # setting max torque and velocity explicitly, overriding the standard parsing of the urdf
        for i, joint in enumerate(self.actuatedJointIds):
            pyb.resetJointState(self.robotId, joint,
                                targetValue = xs[0][i],
                                targetVelocity =xs[0][i + nq])
            pyb.changeDynamics(self.robotId, joint,
                                maxJointVelocity = max_vel[i])
            print(max_vel[i])
            pyb.enableJointForceTorqueSensor(self.robotId, joint)`
The values in the cycle are passed correctly, but just the last one has some effect on the joint dynamics which is clamped at a value 8.13 as you can see in the attached picture (the dashed trajectories are the joint velocities from the simulation).
Why is this bug happening? Shouldn't the function change one joint dynamics at a time? Is there any other option or trick to make this change in the dynamics work as I intend it to?
![bug_velocity](https://user-images.githubusercontent.com/56651708/133809083-d96d4e03-37ca-46cf-a5ca-d3c5da43f291.png)

**Comment by bttner** [comment]
After reading the relevant parts in the server command processor ([search for _m_maxJointVelocity_](https://github.com/bulletphysics/bullet3/blob/master/examples/SharedMemory/PhysicsServerCommandProcessor.cpp)), I am afraid that you can only set the maximum velocity of all links but not individually.  However, when using `setJointMotorControl2` you can define the maximum applied force. Maybe that helps you.

**Comment by gfadini** [comment]
Hi @bttner, thanks for the reply.
I may try with this method, by the way I was already limiting the torques with a saturation in my controller.
Surprisingly I found that still too much energy was being injected into the system, so I tried also to cap the joint speed as a quick fix.
I'm not sure if bullet is the best option for pure torque control

**Comment by erwincoumans** [comment, marked as answer]
The maxJointVelocity in changeDynamics is a safety measure, set to 100 radians per second by default. Your robot should never reach this  safety velocity clamping. The API is there, so you can increase the values, in case you hit this threshold. I'll add a mention in the PyBullet Quickstart Guide.

For setJointMotorControl2, when using position control or velocity control, you can have a per-joint max velocity threshold, once the velocity exceeds no more torque will be added by the actuator. Since you use torque control, this won't help you.

We used PyBullet using pure torque control in our MPC, or feed the torques using external PD control, that should work fine.

Usually energy is dissipated using joint friction/damping (which you can simulate using a velocity motor, with a small maximum (friction) force. This works alongside torque control.

**Comment by gfadini** [comment]
Hi @erwincoumans thanks for clearing out my misunderstanding. Following your advice I will then try to use the torque control and modify a the actuation model by including some additional joint friction to dissipate more energy.

**Comment by gfadini** [comment]
Hi @erwincoumans as a follow up I report a much better behavior following your advice and including a proper model of the actuation of the system. The plot that I have are much closer to the desired trajectories than the one previously shown and the difference mainly comes from the discrepancy between the presence of friction in simulation which unfortunately the trajectory optimizer is not taking into account in the dynamics.
![follow_up](https://user-images.githubusercontent.com/56651708/134314200-3425ec4f-27bc-4314-a05c-e5eb1475f02c.png)
